uint256: Remove unnecessary crypto/common.h use

kallewoof opened the pull request:
Alternative approach: #13258.

There is no reason for `uint256::GetCheapHash()` to support multi-platform endian-ness, since the cheap hash is never used outside of the internal system (it is used in `CAddrInfo` bucket stuff, and that's it). We can remove an entire `#include` and probably some overhead as well. This will remove a number of chained dependencies from `uint256` (`crypto/common.h`, `compat/endian.h`, and `compat/byteswap.h`).

Edit: due to the dependency on little endian in the *tests* for the address manager, #7078 switched to the `ReadLE64()` function. This test reverts that decision and updates the tests instead to work with big and little endian. These tests need to be rethought, possibly removed, as they test basically nothing right now*, but in the meantime, this will address the test errors for big endian.

(* they test that after adding a, b, and c to a bucket, adding d will cause a collision, but adding e will not...)

kallewoof committed:
uint256: Remove unnecessary crypto/common.h use

Empact commented:
For reference, added in #7078, 0.12.0

kallewoof commented:
@Empact Thanks! Wow, this was added because of false positive test failures? I don't know if I can agree with that decision. Hard to test a fix though (no PPC/MIPS here).

sipa commented:
NACK, this is invalid C++. You cannot take a pointer to an object of type A and then dereference it as a pointer of type B, unless the types are compatible, or B is a char type.

This will likely work on little-endian systems, but definitely won't on big-endian code. And even on little-endian systems it is undefined behaviour.

kallewoof commented:
@sipa ea1d189

kallewoof committed:
test: addrman: Track endian based collision points

The address manager tests currently depend on little endianness because they are written with assumptions about collisions occurring at specific
points. This should be revisited, and these tests should potentially be removed as they serve little purpose now.
For now, this commit adds support for big endian assumptions and determines the endianness of the system, basing assumptions on that.

kallewoof commented:
There is an alternative approach to this in #13258.
